assets/__tests__/rendered_test.js

```diff
@@ -0,0 +1,191 @@
+import { Rendered } from '../js/phoenix_live_view'
+
+describe('Rendered', () => {
+  describe('mergeDiff', () => {
+    it('recursively merges two diffs', () => {
+      expect(Rendered.mergeDiff(simpleDiff1, simpleDiff2)).toEqual(
+        simpleDiffResult
+      )
+      expect(Rendered.mergeDiff(deepDiff1, deepDiff2)).toEqual(deepDiffResult)
+    })
+    it('returns the latter diff if it contains a `static` key', () => {
+      const diff1 = { 0: ['a'], 1: ['b'] }
+      const diff2 = { 0: ['c'], static: 'c' }
+      expect(Rendered.mergeDiff(diff1, diff2)).toEqual(diff2)
+    })
+  })
+
+  describe('isNewFingerprint', () => {
+    it('returns true if `diff.static` is truthy', () => {
+      const diff = { static: ['<h2>'] }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(true)
+    })
+    it('returns false if `diff.static` is falsy', () => {
+      const diff = { static: undefined }
+      expect(Rendered.isNewFingerprint(diff)).toEqual(false)
+    })
+    it('returns false if `diff` is undefined', () => {
+      expect(Rendered.isNewFingerprint()).toEqual(false)
+    })
+  })
+
+  describe('toString', () => {
+    it('stringifies a diff', () => {
+      expect(Rendered.toString(simpleDiffResult).trim()).toEqual(
+        `
+<div class="thermostat">
```

Comment:
Oh I was referring to this.  How do you think it should be indented?

Reply:
Or maybe that isn't possible b/c of equality....

Reply:
^ That. Template string literals preserve all whitespace.